cirq-core/cirq/devices/device.py

```diff
@@ -46,6 +50,39 @@ def qubit_set(self) -> Optional[AbstractSet['cirq.Qid']]:
         # Default to the qubits being unknown.
         return None
 
+    def edge_set(self) -> Optional[FrozenSet['cirq.QidPair']]:
+        """Returns a set of qubit edges on the device, if possible.
+
+        This property can be overridden in child classes for special handling.
+        The default handling is: GridQids and LineQids will have neighbors as
+        edges, and others will be fully connected.
+
+        Returns:
+            If the device has a finite set of qubits, then a set of all edges
+            on the device is returned.
+
+            If the device has no well defined finite set of qubits (e.g.
+            `cirq.UnconstrainedDevice` has this property), then `None` is
+            returned.
+        """
+        qs = self.qubit_set()
+        if qs is None:
+            return None
+        if all(isinstance(q, _BaseGridQid) for q in qs):
+            return frozenset(
+                [
+                    QidPair(q, q2)
+                    for q in [cast(_BaseGridQid, q) for q in qs]
+                    for q2 in [q + (0, 1), q + (1, 0)]
+                    if q2 in qs
+                ]
+            )
+        if all(isinstance(q, _BaseLineQid) for q in qs):
```

Comment:
This part potentially will need refactoring when we break out different Device classes e.g. `GridDevice`, `LineDevice`, etc. But for now this should be fine.